pkg/template/jsonpath_test.go

```diff
@@ -0,0 +1,144 @@
+package template_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/triggers/pkg/template"
+)
+
+var objects = `{"a":"v","c":{"d":"e"},"empty": "","null": null, "number": 42}`
+var arrays = `[{"a": "b"}, {"c": "d"}, {"e": "f"}]`
+// Checks that we print JSON strings when the JSONPath selects
+// an array or map value and regular values otherwise
+func TestParseJSONPath(t *testing.T) {
+	var objectBody = fmt.Sprintf(`{"body":%s}`, objects)
+	tests := []struct {
+		name  string
+		expr string
+		in     string
+		want    string
+	}{{
+		name: "objects",
+		in:   objectBody,
+		expr: "$(body)",
+		// TODO: Do we need to escape backslashes for backwards compat?
+		want: objects,
+	}, {
+		name: "array of objects",
+		in: fmt.Sprintf(`{"body":%s}`, arrays),
+		expr: "$(body)",
+		want: arrays,
+	}, {
+		name: "array of values",
+		in: `{"body": ["a", "b", "c"]}`,
+		expr: "$(body)",
+		want: 	`["a", "b", "c"]`,
+	}, {
+		name: "string values",
+		in: objectBody,
+		expr: "$(body.a)",
+		want: "v",
+	}, {
+		name: "string values",
+		in: objectBody,
+		expr: "$(body.a)",
+		want: "v",
+	}, {
+		name: "empty string",
+		in: objectBody,
+		expr: "$(body.empty)",
+		want: "",
+	}, {
+		name: "numbers",
+		in: objectBody,
+		expr: "$(body.number)",
+		want: "42",
+	}, {
+		name: "booleans",
+		in: `{"body": {"bool": true}}`,
+		expr: "$(body.bool)",
+		want: "true",
+	}, {
+		name: "null values",
+		in: objectBody,
+		expr: "$(body.null)",
+		want: "null",
```

Comment:
Is this WAI? Even though the payload says `null` we should probably make a distinction between `null` and `"null"`. I say either default to `""` or have ParseJSONPath return `*string` so we can represent this nil value (my guess is `""` is probably sufficient)

Reply:
Yeah, so I went back and forth on this a few times before settling on "null". I agree that if we were talking about return go values, the correct way should be `*string`. However, for our use case here (i.e. concatenating the json values in string in the trigger templates), that would result in the string `<nil>`.  If we use "" to represent the printable value of `null`, we have no way to differentiate between `null` and the empty string "".

The `"null"` and `null` collision is unfortunate though I think this is the least bad way at the moment. Also, this issue is only valid if the JSONPath selects a field whose value is `null`. If the selection is an object or an array which contains both the string null as well as a null value, we get the following: `{"stringNull":"null","nullValue":null}`